consumer_group.go

```diff
@@ -0,0 +1,405 @@
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return coordinator.SyncGroup(req)
+}
+
+func (c *consumerGroup) balance(members map[string]ConsumerGroupMemberMetadata) (BalanceStrategyPlan, error) {
+	topics := make(map[string][]int32)
+	for _, meta := range members {
+		for _, topic := range meta.Topics {
+			topics[topic] = nil
+		}
+	}
+
+	for topic := range topics {
+		partitions, err := c.client.Partitions(topic)
+		if err != nil {
+			return nil, err
+		}
+		topics[topic] = partitions
+	}
+
+	strategy := c.config.Consumer.Group.Rebalance.Strategy
+	return strategy.Plan(members, topics)
+}
+
+func (c *consumerGroup) heartbeat(generationID int32) error {
+	coordinator, err := c.client.Coordinator(c.groupID)
+	if err != nil {
+		return err
+	}
+
+	resp, err := coordinator.Heartbeat(&HeartbeatRequest{
+		GroupId:      c.groupID,
+		MemberId:     c.memberID,
+		GenerationId: generationID,
+	})
+	if err != nil {
+		_ = coordinator.Close()
+		return err
+	}
+
+	switch resp.Err {
+	case ErrNoError:
+		return nil
+	default:
+		return resp.Err
+	}
+}
+
+// Leaves the cluster, called by Close, protected by lock.
+func (c *consumerGroup) leave() error {
+	if c.memberID == "" {
+		return nil
+	}
+
+	coordinator, err := c.client.Coordinator(c.groupID)
+	if err != nil {
+		return err
+	}
+
+	resp, err := coordinator.LeaveGroup(&LeaveGroupRequest{
+		GroupId:  c.groupID,
+		MemberId: c.memberID,
+	})
+	if err != nil {
+		_ = coordinator.Close()
+		return err
+	}
+
+	// Unset memberID
+	c.memberID = ""
+
+	// Check response
+	switch resp.Err {
+	case ErrRebalanceInProgress, ErrUnknownMemberId, ErrNoError:
+		return nil
+	default:
+		return resp.Err
+	}
+}
+
+// --------------------------------------------------------------------
+
+// ConsumerGroupSession represents a consumer group member session.
+type ConsumerGroupSession interface {
+	// Claims returns the claimed partitions by topic.
+	Claims() map[string][]int32
+	// MemberID returns the cluster member ID.
+	MemberID() string
+	// GenerationID returns the current generation ID.
+	GenerationID() int32
+	// Done exposes a notification channel to signal the end of the session
+	// and to indicate that a new rebalance cycle is due. The session will
+	// continue to hold existing claims until Close() is called, to allow
+	// all tasks to flush any pending data and commit offsets.
+	// Please note that brokers will wait for no more than
+	// Consumer.Group.Rebalance.Timeout before evicting the worker from
+	// the group.
+	Done() <-chan struct{}
+	// Release aborts the session and releases the claims.
+	Release() error
+}
+
+type consumerGroupSession struct {
+	claims map[string][]int32
+	parent *consumerGroup
+
+	generationID int32
+
+	closing     chan none
+	done        chan struct{}
+	lastError   error
+	releaseOnce sync.Once
+}
+
+func (s *consumerGroupSession) Claims() map[string][]int32 { return s.claims }
+func (s *consumerGroupSession) MemberID() string           { return s.parent.memberID }
```

Comment:
should this just be exposed on the ConsumerGroup itself?

Reply:
It's not really helpful as these are only available for the duration of the session. A session can be terminated by the server (failing heartbeat) or by the client (heartbeat is stopped)